src/NuGet.Core/NuGet.ProjectModel/MacroStringsUtility.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.ProjectModel
+{
+    // Add JsonPackageReader and PackageSpecWriter tests.
+    internal static class MacroStringsUtility
+    {
+        internal const string NUGET_ENABLE_EXPERIMENTAL_MACROS = nameof(NUGET_ENABLE_EXPERIMENTAL_MACROS);
+        internal const string UserMacro = "$(User)";
+
+        /// <summary>
+        /// Applies macros in place to every string in the list.
+        /// Macros are applied only at the beginning of a string.
+        /// </summary>
+        /// <param name="list">The list of elements to apply the macro on.</param>
+        /// <param name="macroValue">The macro value that'll need to match from the strings.</param>
+        /// <param name="macroName">The macro that'll replace the value within the string.</param>
+        /// <param name="stringComparison">The comparer to use. Normally these strings are paths, so the comparer should be OS aware.</param>
+        internal static void ApplyMacros(IList<string> list, string macroValue, string macroName, StringComparison stringComparison)
+        {
+            if (list != null)
+            {
+                for (int i = 0; i < list.Count; i++)
+                {
+                    var result = ApplyMacro(list[i], macroValue, macroName, stringComparison);
+                    list[i] = result;
+                }
+            }
+        }
+
+        internal static void ExtractMacros(List<string> list, string macroValue, string macroName)
+        {
+            if (list != null)
+            {
+                for (int i = 0; i < list.Count; i++)
+                {
+                    var result = ExtractMacro(list[i], macroValue, macroName);
+                    list[i] = result;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Applies macro to the string in the list.
+        /// </summary>
+        /// <param name="originalString">The string to apply the macro on.</param>
+        /// <param name="macroValue">The macro value that'll need to match from the string.</param>
+        /// <param name="macroName">The macro that'll replace the value within the string.</param>
+        /// <param name="stringComparison">The comparer to use. Normally these strings are paths, so the comparer should be OS aware.</param>
+        internal static string ApplyMacro(string originalString, string macroValue, string macroName, StringComparison stringComparison)
+        {
+            if (!string.IsNullOrEmpty(originalString) && !string.IsNullOrEmpty(macroValue) && originalString.StartsWith(macroValue, stringComparison))
+            {
+                return macroName + originalString.Substring(macroValue.Length);
```

Comment:
Has the perf of this been measured?  For example, on OrchardCore?  I hope it's immeasurable, but I guess it's better to validate than be surprised.

I'm not how how frequently paths are used in the assets file, but both here and in `ExtractMacro` there are 2 additional string allocations per call (that use the macro). NuGet will need to do these 2 additional allocations per path on write, and then again on read, so effectively 4 additional allocations per affected path on every CI restore + build operation.

My concern is that this PR's feature only benefits the first restore on a devbox. And in this case, the global packages folder is already filled, so it shouldn't be a long restore. But the perf impact is on every single restore, on every single machine, including on CI.

Given we're frequently asked to improve restore performance, I'm wondering if this feature is conflicting with our perf requirements.

Perhaps the question is, is the environment variable to enable the feature intended to be a temporary thing, while we test that it doesn't break anyone, or do we intend on keeping this environment variable forever (and keep the feature off by default forever), so that only devboxes can opt-in to the feature?  In this case, the PR checklist might be wrong, because no documentation PR is filed to update the environment variables list doc page.

Reply:
I tried measuring, but no difference.

> I'm not how how frequently paths are used in the assets file, but both here and in ExtractMacro there are 2 additional string allocations per call (that use the macro). NuGet will need to do these 2 additional allocations per path on write, and then again on read, so effectively 4 additional allocations per affected path on every CI restore + build operation.

That's why I minimized the number of changes because it'll almost always be like 2-3 paths per project, which is not a crazy amount. 

The thing with devbox is that the priority is fast "get to code", so the cost of 6 allocations is better than the cost of project-system redoing their change because nuget touched the assets file with a change that doesn't really change anything for the project system (basically saying a new config was added).

Reply:
> The thing with devbox is that the priority is fast "get to code"

I understand, but the point I was trying to make is that some optimizations to make "first open solution and starting coding" can be harmful to "this happens on every CI build, where there is no user interactively waiting for anything".

However, if the change doesn't impact perf in a measurable way, and especially if we keep a feature flag which we except never to be used on CI, then it could be fine.